### PR TITLE
add KifiSourceAttribution

### PR DESCRIPTION
**Pull request by camhashemi**
@leogrim @ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala [hunk 1]
```diff
@@ -50,9 +51,11 @@ object KeepFields {
   object Source {
     def apply(teamId: SlackTeamId, channelId: SlackChannelId): String = s"slack|${channelId.value}" // TODO(Léo): this should be "slack|teamId_channelId"
     def apply(handle: TwitterHandle): String = s"twitter|${handle.value}"
+    def apply(user: BasicUser): String = s"kifi|${user.externalId.id}"
     def apply(source: SourceAttribution): String = source match {
       case TwitterAttribution(tweet) => Source(tweet.user.screenName)
       case SlackAttribution(message, teamId) => Source(teamId, message.channel.id)
+      case KifiAttribution(keptBy, _, _, _) => Source(keptBy) // todo(cam): ask Léo what we should do here
```
Comment (camhashemi) [inline kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala]: @leogrim what should we do here?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/gen/KeepActivityGen.scala [hunk 2]
```diff
@@ -50,7 +50,7 @@ object KeepActivityGen {
           }
       }
 
-      val source = sourceAttrOpt.map(_._1).map {
+      val source = sourceAttrOpt.map(_._1).collect {
         case SlackAttribution(message, _) => KeepEventSource(KeepEventSourceKind.Slack, Some(message.permalink))
         case TwitterAttribution(tweet) => KeepEventSource(KeepEventSourceKind.Twitter, Some(tweet.permalink))
       }
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/gen/KeepActivityGen.scala:56]: @ryanpbrewster this is what I meant about "third party attribution" versus "kifi attribution", "kifi attribution" is handled by default, usually `SourceAttribution` adds some extra info to a feature.

File: kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala]: Almost this, use the internal `userId` instead.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala [hunk 3]
```diff
@@ -192,10 +192,14 @@ class KeepInternerImpl @Inject() (
...
-  private def internKeep(uri: NormalizedURI, userIdOpt: Option[Id[User]], libraryOpt: Option[Library], source: KeepSource,
+  private def internKeep(uri: NormalizedURI, userIdOpt: Option[Id[User]], libraryOpt: Option[Library], usersAdded: Set[Id[User]], source: KeepSource,
     title: Option[String], url: String, keptAt: DateTime,
-    sourceAttribution: Option[RawSourceAttribution], note: Option[String])(implicit session: RWSession) = {
-    airbrake.verify(userIdOpt.isDefined || sourceAttribution.isDefined, s"interning a keep (uri ${uri.id.get}, lib ${libraryOpt.map(_.id.get)}) with no user AND no source?!?!?!")
+    thirdPartyAttribution: Option[RawSourceAttribution], note: Option[String])(implicit session: RWSession) = {
+    airbrake.verify(userIdOpt.isDefined || thirdPartyAttribution.isDefined, s"interning a keep (uri ${uri.id.get}, lib ${libraryOpt.map(_.id.get)}) with no user AND no source?!?!?!")
+
+    val sourceAttribution = thirdPartyAttribution.orElse {
+      userIdOpt.map(userId => RawKifiAttribution(userId, usersAdded, Set.empty, libraryOpt.map(_.id.get).toSet))
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: Is this temporary? I don't know whether we want the `libraryOpt` in here cc @ryanpbrewster

File: kifi-backend/modules/common/app/com/keepit/model/KeepSourceAttribution.scala [hunk 4]
```diff
@@ -84,3 +88,8 @@ case class RawSlackAttribution(message: SlackMessage, teamId: SlackTeamId) exten
 object RawSlackAttribution {
   implicit val format = Json.format[RawSlackAttribution]
 }
+
+case class RawKifiAttribution(author: Id[User], users: Set[Id[User]], nonUsers: Set[EmailAddress], libraries: Set[Id[Library]]) extends RawSourceAttribution
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/KeepSourceAttribution.scala]: Have `initialRecipients` field of type `KeepConnections`, to be consistent with `Keep`

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/client/ClientKeepMutationController.scala [hunk 5]
```diff
@@ -43,7 +43,7 @@ class ClientKeepMutationController @Inject() (
         author = Author.KifiUser(request.userId),
         url = externalCreateRequest.url,
         source = externalCreateRequest.source,
-        attribution = None,
+        attribution = RawKifiAttribution(request.userId, users, nonUsers = Set.empty, libraries),
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/client/ClientKeepMutationController.scala]: this can be `externalCreateRequest.emails` now

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/client/ClientKeepMutationController.scala]: niceee

File: kifi-backend/modules/common/app/com/keepit/model/KeepSourceAttribution.scala (hunk 4 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/KeepSourceAttribution.scala]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala (hunk 3 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: the goal is to persist whatever libraries are connected to that keep upon creation

File: kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/search/app/com/keepit/search/index/graph/keep/KeepIndexable.scala]: talked in meatspace, we should filter out `KifiAttributions` in `SearchControllerUtil.getSources`

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala (hunk 3 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: I'm planning on replacing the existing `internRawBookmarksWithStatus` calls with a simpler `internKeep` (see above) call. All of this code is going to go away soon.

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: yeah I added this logic there too, just wanted to make sure we were doing the right thing in the interim.

**Comment by camhashemi** [comment]
thanks guys. going to see if things are working in prod before running the migration.

**Pull request merged**